### PR TITLE
cellRtc: some improvements

### PR DESCRIPTION
**Pull request by Megamouse**
- implement Rfc3339 part of cellRtcParseDateTime
- simplify cellRtcParseRfc3339
- return a value in cellRtcGetCurrentSecureTick

### CODE REVIEW

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 1]
```diff
@@ -488,12 +488,34 @@ error_code cellRtcFormatRfc3339LocalTime(vm::ptr<char> pszDateTime, vm::cptr<Cel
...
+	cellRtc.todo("cellRtcParseDateTime(pUtc=*0x%x, pszDateTime=%s)", pUtc, pszDateTime);
+
+	if (!vm::check_addr(pUtc.addr()) || !vm::check_addr(pszDateTime.addr()))
+	{
+		return CELL_RTC_ERROR_INVALID_POINTER;
+	}
+
+	u32 pos = 0;
+
+	while (pszDateTime[pos] == 32 || pszDateTime[pos] == 9)
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:507]: You sure there is no another terminator here? seems odd.

Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellRtc.cpp:507]: It's skipping spaces and tabs. every other character breaks the loop

Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellRtc.cpp:507]: it's also done later again. but I didn't want to implement the rest right now

**Comment by capriots** [comment]
I actually implemented cellRtc ~90% a year ago or so but didn't have time to finish it and just kinda forgot about it.
https://github.com/capriots/rpcs3/tree/cellRtc
I can review this PR in a day or two. I should have time to finish my own implementation next month.

Also here is my Ghidra export, maybe this is useful for you (doesn't include the actual binary, don't worry):
[librtc.prx.xml.zip](https://github.com/RPCS3/rpcs3/files/14893874/librtc.prx.xml.zip)

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp (hunk 1 above)
Comment (capriots) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:507]:
```suggestion
	while (std::isblank(pszDateTime[pos]))
```

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 2]
```diff
@@ -571,20 +594,15 @@ error_code cellRtcParseRfc3339(vm::ptr<CellRtcTick> pUtc, vm::cptr<char> pszDate
...
-	// TODO fix unholyness
-
-	s64 l_add;
-	u32 pc_var_5;
-	u32 u_var_7;
-	s32 i_var_6;
-	s32 i_var_8;
+	const auto digit = [](char c) -> s32
+	{
+		return c - 48;
```
Comment (capriots) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:599]:
```suggestion
		return c - '0';
```
I personally prefer this, but it should be clear in this context

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 3]
```diff
@@ -1189,6 +1187,8 @@ error_code cellRtcGetCurrentSecureTick(vm::ptr<CellRtcTick> tick)
 
 	// TODO
 
+	tick->tick = 0xe01d003a63a000;
```
Comment (capriots) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:1190]: You can use cellRtcGetCurrentTick() for now if you want.
The syscall here returns the hardware clock corrected for drift. It is unaffected by the date and time settings in the XMB/VSH settings, which makes it "secure" I guess.

**Comment by capriots** [comment]
I went over the decompilation for cellRtcParseRfc3339() again and your implementation seems correct. You can actually change the log level to notice, as there is nothing more to add I believe.

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp [hunk 4]
```diff
@@ -620,124 +644,98 @@ error_code cellRtcParseRfc3339(vm::ptr<CellRtcTick> pUtc, vm::cptr<char> pszDate
...
+		if (sign == '-')
 		{
-			l_add = i_var_8;
+			minutes_to_add = -minutes_to_add;
 		}
 	}
 
 	cellRtcGetTick(date_time, pUtc);
-	cellRtcTickAddMinutes(pUtc, pUtc, l_add);
+	cellRtcTickAddMinutes(pUtc, pUtc, minutes_to_add);
```
Comment (capriots) [inline rpcs3/Emu/Cell/Modules/cellRtc.cpp:738]:
```suggestion
	cellRtcTickAddMinutes(pUtc, pUtc, -minutes_to_add);
```
We need to _remove_ the time zone adjustment, the preceding string is already in local time and the function always returns UTC.

Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellRtc.cpp:738]: The sign is already applied in the previous scope though

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp (hunk 3 above)
Comment (Megamouse) [reply rpcs3/Emu/Cell/Modules/cellRtc.cpp:1190]: I'm not gonna add stuff I haven't personally confirmed and I'm not gonna look into this any more for now.

File: rpcs3/Emu/Cell/Modules/cellRtc.cpp (hunk 4 above)
Comment (capriots) [reply rpcs3/Emu/Cell/Modules/cellRtc.cpp:738]: Wait, am I stupid? You need to invert the sign, so for example if the string is "+01:00" then you need to subtract 60 minutes from the final tick and if it's "-01:00" you need to add 60 minutes.

**Pull request merged**